ICE: compiler/rustc_middle/src/ty/context.rs:705:13: node_type: no type for node

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
extern crate rustc_codegen_ssa;

use rustc_codegen_ssa::CodegenResults;
use rustc_codegen_ssa::traits::{CodegenBackend, WriteBackendMethods};

#[derive(Clone)]
pub struct GccCodegenBackend;

impl CodegenBackend for GccCodegenBackend {
    fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
        use rustc_codegen_ssa::back::link::link_binary;

        link_binary::<crate::archive::ArArchiveBuilder<'_>>(
            sess,
            &codegen_results,
            outputs,
        )
    }
}

impl WriteBackendMethods for GccCodegenBackend {}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (1b57946a4 2022-08-03)
binary: rustc
commit-hash: 1b57946a405d5b2a87e612335db033edb2c3427f
commit-date: 2022-08-03
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### Error output

```error[E0433]: failed to resolve: unresolved import
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:13:30
   |
13 |         link_binary::<crate::archive::ArArchiveBuilder<'_>>(
   |                              ^^^^^^^
   |                              |
   |                              unresolved import
   |                              help: a similar path exists: `rustc_codegen_ssa::back::archive`

error[E0412]: cannot find type `Session` in this scope
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:10:27
   |
10 |     fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
   |                           ^^^^^^^ not found in this scope

error[E0412]: cannot find type `OutputFilenames` in this scope
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:10:79
   |
10 |     fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
   |                                                                               ^^^^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ErrorReported` in this scope
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:10:110
   |
9  | impl CodegenBackend for GccCodegenBackend {
   |     - help: you might be missing a type parameter: `<ErrorReported>`
10 |     fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
   |                                                                                                              ^^^^^^^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `9b62bf4183779f002aba901cbe888acec64ea7c5`
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:21:50
   |
21 | impl WriteBackendMethods for GccCodegenBackend {}
   |                                                  ^ consider adding a `main` function to `9b62bf4183779f002aba901cbe888acec64ea7c5.rs`

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:1:1
  |
1 | extern crate rustc_codegen_ssa;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:3:5
  |
3 | use rustc_codegen_ssa::CodegenResults;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:4:33
  |
4 | use rustc_codegen_ssa::traits::{CodegenBackend, WriteBackendMethods};
  |                                 ^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:4:49
  |
4 | use rustc_codegen_ssa::traits::{CodegenBackend, WriteBackendMethods};
  |                                                 ^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:10:5
   |
10 | /     fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
11 | |         use rustc_codegen_ssa::back::link::link_binary;
12 | |
13 | |         link_binary::<crate::archive::ArArchiveBuilder<'_>>(
...  |
17 | |         )
18 | |     }
   | |_____^
   |
   = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
   = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:9:6
  |
9 | impl CodegenBackend for GccCodegenBackend {
  |      ^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:11:13
   |
11 |         use rustc_codegen_ssa::back::link::link_binary;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
   = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:21:6
   |
21 | impl WriteBackendMethods for GccCodegenBackend {}
   |      ^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
   = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:10:53
   |
10 |     fn link(&self, sess: &Session, codegen_results: CodegenResults, outputs: &OutputFilenames) -> Result<(), ErrorReported> {
   |                                                     ^^^^^^^^^^^^^^
   |
   = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
   = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:13:9
   |
13 |         link_binary::<crate::archive::ArArchiveBuilder<'_>>(
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
   = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0046]: not all trait items implemented, missing: `codegen_crate`, `join_codegen`
 --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:9:1
  |
9 | impl CodegenBackend for GccCodegenBackend {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `codegen_crate`, `join_codegen` in implementation
  |
  = help: implement the missing item: `fn codegen_crate(&self, _: rustc_middle::ty::context::TyCtxt<'tcx>, _: rustc_metadata::rmeta::encoder::EncodedMetadata, _: bool) -> Box<(dyn Any + 'static)> { todo!() }`
  = help: implement the missing item: `fn join_codegen(&self, _: Box<(dyn Any + 'static)>, _: &rustc_session::session::Session, _: &rustc_session::config::OutputFilenames) -> Result<(CodegenResults, HashMap<rustc_query_system::dep_graph::dep_node::WorkProductId, rustc_query_system::dep_graph::graph::WorkProduct, BuildHasherDefault<rustc_hash::FxHasher>>), rustc_errors::ErrorGuaranteed> { todo!() }`

error[E0046]: not all trait items implemented, missing: `Module`, `TargetMachine`, `ModuleBuffer`, `Context`, `ThinData`, `ThinBuffer`, `run_link`, `run_fat_lto`, `run_thin_lto`, `print_pass_timings`, `optimize`, `optimize_fat`, `optimize_thin`, `codegen`, `prepare_thin`, `serialize_module`
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:21:1
   |
21 | impl WriteBackendMethods for GccCodegenBackend {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Module`, `TargetMachine`, `ModuleBuffer`, `Context`, `ThinData`, `ThinBuffer`, `run_link`, `run_fat_lto`, `run_thin_lto`, `print_pass_timings`, `optimize`, `optimize_fat`, `optimize_thin`, `codegen`, `prepare_thin`, `serialize_module` in implementation
   |
   = help: implement the missing item: `type Module = Type;`
   = help: implement the missing item: `type TargetMachine = Type;`
   = help: implement the missing item: `type ModuleBuffer = Type;`
   = help: implement the missing item: `type Context = Type;`
   = help: implement the missing item: `type ThinData = Type;`
   = help: implement the missing item: `type ThinBuffer = Type;`
   = help: implement the missing item: `fn run_link(_: &CodegenContext<Self>, _: &rustc_errors::Handler, _: Vec<ModuleCodegen<<Self as WriteBackendMethods>::Module>>) -> Result<ModuleCodegen<<Self as WriteBackendMethods>::Module>, rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `fn run_fat_lto(_: &CodegenContext<Self>, _: Vec<FatLTOInput<Self>>, _: Vec<(SerializedModule<<Self as WriteBackendMethods>::ModuleBuffer>, rustc_query_system::dep_graph::graph::WorkProduct)>) -> Result<LtoModuleCodegen<Self>, rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `fn run_thin_lto(_: &CodegenContext<Self>, _: Vec<(String, <Self as WriteBackendMethods>::ThinBuffer)>, _: Vec<(SerializedModule<<Self as WriteBackendMethods>::ModuleBuffer>, rustc_query_system::dep_graph::graph::WorkProduct)>) -> Result<(Vec<LtoModuleCodegen<Self>>, Vec<rustc_query_system::dep_graph::graph::WorkProduct>), rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `fn print_pass_timings(&self) { todo!() }`
   = help: implement the missing item: `unsafe fn optimize(_: &CodegenContext<Self>, _: &rustc_errors::Handler, _: &ModuleCodegen<<Self as WriteBackendMethods>::Module>, _: &ModuleConfig) -> Result<(), rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `fn optimize_fat(_: &CodegenContext<Self>, _: &mut ModuleCodegen<<Self as WriteBackendMethods>::Module>) -> Result<(), rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `unsafe fn optimize_thin(_: &CodegenContext<Self>, _: ThinModule<Self>) -> Result<ModuleCodegen<<Self as WriteBackendMethods>::Module>, rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `unsafe fn codegen(_: &CodegenContext<Self>, _: &rustc_errors::Handler, _: ModuleCodegen<<Self as WriteBackendMethods>::Module>, _: &ModuleConfig) -> Result<CompiledModule, rustc_span::fatal_error::FatalError> { todo!() }`
   = help: implement the missing item: `fn prepare_thin(_: ModuleCodegen<<Self as WriteBackendMethods>::Module>) -> (String, <Self as WriteBackendMethods>::ThinBuffer) { todo!() }`
   = help: implement the missing item: `fn serialize_module(_: ModuleCodegen<<Self as WriteBackendMethods>::Module>) -> (String, <Self as WriteBackendMethods>::ModuleBuffer) { todo!() }`

error[E0107]: this function takes 0 generic arguments but 1 generic argument was supplied
  --> 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:13:9
   |
13 |         link_binary::<crate::archive::ArArchiveBuilder<'_>>(
   |         ^^^^^^^^^^^---------------------------------------- help: remove these generics
   |         |
   |         expected 0 generic arguments

```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_middle/src/ty/context.rs:705:13: node_type: no type for node `type crate::archive::ArArchiveBuilder<'_> (hir_id=HirId { owner: DefId(0:15 ~ 9b62bf4183779f002aba901cbe888acec64ea7c5[15ed]::{impl#0}::link), local_id: 10 })`

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/compiler/rustc_errors/src/lib.rs:1392:9
stack backtrace:
   0:     0x7f6d529ca280 - std::backtrace_rs::backtrace::libunwind::trace::h9d4b0a0070383433
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   1:     0x7f6d529ca280 - std::backtrace_rs::backtrace::trace_unsynchronized::hd0319debd80d7567
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f6d529ca280 - std::sys_common::backtrace::_print_fmt::ha358f64ddabcaade
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f6d529ca280 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbb8bd1ad8e501cae
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f6d52a25c4c - core::fmt::write::hba0660bcdf7a814a
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/core/src/fmt/mod.rs:1198:17
   5:     0x7f6d529bb2c5 - std::io::Write::write_fmt::h1d8458299a37d708
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/io/mod.rs:1672:15
   6:     0x7f6d529ccf61 - std::sys_common::backtrace::_print::hc6c01539c480a60d
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f6d529ccf61 - std::sys_common::backtrace::print::hb6a72c6c5063f3fd
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f6d529ccf61 - std::panicking::default_hook::{{closure}}::hb421182db9227a0c
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/panicking.rs:295:22
   9:     0x7f6d529ccc2e - std::panicking::default_hook::h57cf011ec63882a8
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/panicking.rs:314:9
  10:     0x7f6d5525d5d4 - rustc_driver[28a8a24145fee7e3]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f6d529cd796 - std::panicking::rust_panic_with_hook::h93256441b3a09185
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/panicking.rs:702:17
  12:     0x7f6d561fbed1 - std[1797f99dfd5db856]::panicking::begin_panic::<rustc_errors[57ee69a351dd2f5f]::ExplicitBug>::{closure#0}
  13:     0x7f6d561fb3c6 - std[1797f99dfd5db856]::sys_common::backtrace::__rust_end_short_backtrace::<std[1797f99dfd5db856]::panicking::begin_panic<rustc_errors[57ee69a351dd2f5f]::ExplicitBug>::{closure#0}, !>
  14:     0x7f6d561dc416 - std[1797f99dfd5db856]::panicking::begin_panic::<rustc_errors[57ee69a351dd2f5f]::ExplicitBug>
  15:     0x7f6d56220956 - std[1797f99dfd5db856]::panic::panic_any::<rustc_errors[57ee69a351dd2f5f]::ExplicitBug>
  16:     0x7f6d562207a5 - <rustc_errors[57ee69a351dd2f5f]::HandlerInner>::bug::<&alloc[56eb4d0e39078a]::string::String>
  17:     0x7f6d56220500 - <rustc_errors[57ee69a351dd2f5f]::Handler>::bug::<&alloc[56eb4d0e39078a]::string::String>
  18:     0x7f6d5624341d - rustc_middle[3c62e2e09efd4a65]::ty::context::tls::with_context_opt::<rustc_middle[3c62e2e09efd4a65]::ty::context::tls::with_opt<rustc_middle[3c62e2e09efd4a65]::util::bug::opt_span_bug_fmt<rustc_span[35b28c3ec96f8264]::span_encoding::Span>::{closure#0}, ()>::{closure#0}, ()>
  19:     0x7f6d562464d6 - rustc_middle[3c62e2e09efd4a65]::util::bug::opt_span_bug_fmt::<rustc_span[35b28c3ec96f8264]::span_encoding::Span>
  20:     0x7f6d53cb6823 - rustc_middle[3c62e2e09efd4a65]::util::bug::bug_fmt
  21:     0x7f6d53bc7b7a - <rustc_middle[3c62e2e09efd4a65]::ty::context::TypeckResults>::node_type
  22:     0x7f6d5559a454 - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::point_at_type_arg_instead_of_call_if_possible
  23:     0x7f6d53cc67ac - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::check_argument_types
  24:     0x7f6d53c7b167 - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::check_call
  25:     0x7f6d53cc95af - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  26:     0x7f6d53cdf08d - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::check_block_with_expected
  27:     0x7f6d53cc971e - <rustc_typeck[310de37840dc46b8]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  28:     0x7f6d54144e5d - rustc_typeck[310de37840dc46b8]::check::check::check_fn
  29:     0x7f6d5413700e - <rustc_infer[851aa0b77966ddaa]::infer::InferCtxtBuilder>::enter::<&rustc_middle[3c62e2e09efd4a65]::ty::context::TypeckResults, <rustc_typeck[310de37840dc46b8]::check::inherited::InheritedBuilder>::enter<rustc_typeck[310de37840dc46b8]::check::typeck_with_fallback<rustc_typeck[310de37840dc46b8]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[3c62e2e09efd4a65]::ty::context::TypeckResults>::{closure#0}>
  30:     0x7f6d541348e4 - rustc_typeck[310de37840dc46b8]::check::typeck
  31:     0x7f6d548e0f6f - rustc_query_system[3672fd42db334255]::query::plumbing::try_execute_query::<rustc_query_impl[56619d473d589f9a]::plumbing::QueryCtxt, rustc_query_system[3672fd42db334255]::query::caches::DefaultCache<rustc_span[35b28c3ec96f8264]::def_id::LocalDefId, &rustc_middle[3c62e2e09efd4a65]::ty::context::TypeckResults>>
  32:     0x7f6d5515ffc8 - <rustc_query_impl[56619d473d589f9a]::Queries as rustc_middle[3c62e2e09efd4a65]::ty::query::QueryEngine>::typeck
  33:     0x7f6d54f50c76 - rustc_data_structures[fa69b12c2d991fb7]::sync::par_for_each_in::<&[rustc_span[35b28c3ec96f8264]::def_id::LocalDefId], <rustc_middle[3c62e2e09efd4a65]::hir::map::Map>::par_body_owners<rustc_typeck[310de37840dc46b8]::check::typeck_item_bodies::{closure#0}>::{closure#0}>
  34:     0x7f6d54f50a2b - <rustc_middle[3c62e2e09efd4a65]::hir::map::Map>::par_body_owners::<rustc_typeck[310de37840dc46b8]::check::typeck_item_bodies::{closure#0}>
  35:     0x7f6d54f5097c - rustc_typeck[310de37840dc46b8]::check::typeck_item_bodies
  36:     0x7f6d54f397e5 - rustc_query_system[3672fd42db334255]::query::plumbing::try_execute_query::<rustc_query_impl[56619d473d589f9a]::plumbing::QueryCtxt, rustc_query_system[3672fd42db334255]::query::caches::DefaultCache<(), ()>>
  37:     0x7f6d54f39571 - rustc_query_system[3672fd42db334255]::query::plumbing::get_query::<rustc_query_impl[56619d473d589f9a]::queries::typeck_item_bodies, rustc_query_impl[56619d473d589f9a]::plumbing::QueryCtxt>
  38:     0x7f6d54c286bf - <rustc_session[a98b15f73d1a4597]::session::Session>::time::<(), rustc_typeck[310de37840dc46b8]::check_crate::{closure#7}>
  39:     0x7f6d54c2813f - rustc_typeck[310de37840dc46b8]::check_crate
  40:     0x7f6d54c27747 - rustc_interface[2779bea5779b9cc7]::passes::analysis
  41:     0x7f6d54ff7cf5 - rustc_query_system[3672fd42db334255]::query::plumbing::try_execute_query::<rustc_query_impl[56619d473d589f9a]::plumbing::QueryCtxt, rustc_query_system[3672fd42db334255]::query::caches::DefaultCache<(), core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>>
  42:     0x7f6d54ff7a7e - rustc_query_system[3672fd42db334255]::query::plumbing::get_query::<rustc_query_impl[56619d473d589f9a]::queries::analysis, rustc_query_impl[56619d473d589f9a]::plumbing::QueryCtxt>
  43:     0x7f6d54a39bca - <rustc_interface[2779bea5779b9cc7]::interface::Compiler>::enter::<rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}::{closure#2}, core[4c00d720cb23a2ad]::result::Result<core[4c00d720cb23a2ad]::option::Option<rustc_interface[2779bea5779b9cc7]::queries::Linker>, rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>
  44:     0x7f6d54a2d5fa - rustc_span[35b28c3ec96f8264]::with_source_map::<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_interface[2779bea5779b9cc7]::interface::create_compiler_and_run<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}>::{closure#1}>
  45:     0x7f6d54a2cfd0 - rustc_interface[2779bea5779b9cc7]::interface::create_compiler_and_run::<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}>
  46:     0x7f6d54a2b341 - <scoped_tls[b1559f3fa2997ec4]::ScopedKey<rustc_span[35b28c3ec96f8264]::SessionGlobals>>::set::<rustc_interface[2779bea5779b9cc7]::interface::run_compiler<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}>::{closure#0}, core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>
  47:     0x7f6d54a2b02f - std[1797f99dfd5db856]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[2779bea5779b9cc7]::util::run_in_thread_pool_with_globals<rustc_interface[2779bea5779b9cc7]::interface::run_compiler<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}>::{closure#0}, core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>::{closure#0}, core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>
  48:     0x7f6d55084c49 - <<std[1797f99dfd5db856]::thread::Builder>::spawn_unchecked_<rustc_interface[2779bea5779b9cc7]::util::run_in_thread_pool_with_globals<rustc_interface[2779bea5779b9cc7]::interface::run_compiler<core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>, rustc_driver[28a8a24145fee7e3]::run_compiler::{closure#1}>::{closure#0}, core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>::{closure#0}, core[4c00d720cb23a2ad]::result::Result<(), rustc_errors[57ee69a351dd2f5f]::ErrorGuaranteed>>::{closure#1} as core[4c00d720cb23a2ad]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x7f6d529d7763 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7cdbc89df5cc56b5
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/alloc/src/boxed.rs:1935:9
  50:     0x7f6d529d7763 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h56fd0a7120358d6b
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/alloc/src/boxed.rs:1935:9
  51:     0x7f6d529d7763 - std::sys::unix::thread::Thread::new::thread_start::hab11f20f5d17dbc7
                               at /rustc/1b57946a405d5b2a87e612335db033edb2c3427f/library/std/src/sys/unix/thread.rs:108:17
  52:     0x7f6d5270e54d - <unknown>
  53:     0x7f6d52793874 - clone
  54:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.64.0-nightly (1b57946a4 2022-08-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `<impl at 9b62bf4183779f002aba901cbe888acec64ea7c5.rs:9:1: 9:42>::link`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 19 previous errors

Some errors have detailed explanations: E0046, E0107, E0412, E0433, E0601, E0658.
For more information about an error, try `rustc --explain E0046`.
```

</p>
</details>

--- a/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
@@ -1761,13 +1761,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             .filter_map(|seg| seg.args.as_ref())
                             .flat_map(|a| a.args.iter())
                         {
-                            if let hir::GenericArg::Type(hir_ty) = &arg {
-                                let ty = self.resolve_vars_if_possible(
-                                    self.typeck_results.borrow().node_type(hir_ty.hir_id),
-                                );
-                                if ty == predicate.self_ty() {
-                                    error.obligation.cause.span = hir_ty.span;
-                                }
+                            if let hir::GenericArg::Type(hir_ty) = &arg
+                                && let Some(ty) =
+                                    self.typeck_results.borrow().node_type_opt(hir_ty.hir_id)
+                                && self.resolve_vars_if_possible(ty) == predicate.self_ty()
+                            {
+                                error.obligation.cause.span = hir_ty.span;
+                                break;
                             }
                         }
                     }

--- a/src/test/ui/argument-suggestions/issue-100154.rs
+++ b/src/test/ui/argument-suggestions/issue-100154.rs
@@ -1,0 +1,7 @@
+fn foo(i: impl std::fmt::Display) {}
+
+fn main() {
+    foo::<()>(());
+    //~^ ERROR this function takes 0 generic arguments but 1 generic argument was supplied
+    //~| ERROR `()` doesn't implement `std::fmt::Display`
+}

--- a/src/test/ui/argument-suggestions/issue-100154.stderr
+++ b/src/test/ui/argument-suggestions/issue-100154.stderr
@@ -1,0 +1,35 @@
+error[E0107]: this function takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/issue-100154.rs:4:5
+   |
+LL |     foo::<()>(());
+   |     ^^^------ help: remove these generics
+   |     |
+   |     expected 0 generic arguments
+   |
+note: function defined here, with 0 generic parameters
+  --> $DIR/issue-100154.rs:1:4
+   |
+LL | fn foo(i: impl std::fmt::Display) {}
+   |    ^^^
+   = note: `impl Trait` cannot be explicitly specified as a generic argument
+
+error[E0277]: `()` doesn't implement `std::fmt::Display`
+  --> $DIR/issue-100154.rs:4:15
+   |
+LL |     foo::<()>(());
+   |     --------- ^^ `()` cannot be formatted with the default formatter
+   |     |
+   |     required by a bound introduced by this call
+   |
+   = help: the trait `std::fmt::Display` is not implemented for `()`
+   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
+note: required by a bound in `foo`
+  --> $DIR/issue-100154.rs:1:16
+   |
+LL | fn foo(i: impl std::fmt::Display) {}
+   |                ^^^^^^^^^^^^^^^^^ required by this bound in `foo`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0107, E0277.
+For more information about an error, try `rustc --explain E0107`.

--- a/src/test/ui/transmutability/references.stderr
+++ b/src/test/ui/transmutability/references.stderr
@@ -1,8 +1,8 @@
 error[E0277]: `&'static Unit` cannot be safely transmuted into `&'static Unit` in the defining scope of `assert::Context`.
-  --> $DIR/references.rs:19:52
+  --> $DIR/references.rs:19:37
    |
 LL |     assert::is_maybe_transmutable::<&'static Unit, &'static Unit>();
-   |                                                    ^^^^^^^^^^^^^ `&'static Unit` cannot be safely transmuted into `&'static Unit` in the defining scope of `assert::Context`.
+   |                                     ^^^^^^^^^^^^^ `&'static Unit` cannot be safely transmuted into `&'static Unit` in the defining scope of `assert::Context`.
    |
    = help: the trait `BikeshedIntrinsicFrom<&'static Unit, assert::Context, true, true, true, true>` is not implemented for `&'static Unit`
 note: required by a bound in `is_maybe_transmutable`

